packages/nodes-base/nodes/SplitInBatches/v3/SplitInBatchesV3.node.ts

```diff
@@ -154,11 +197,81 @@ export class SplitInBatchesV3 implements INodeType {
 
 		if (returnItems.length === 0) {
 			nodeContext.done = true;
+			// Normal completion - always reset counter to prevent memory leaks
+			// The infinite loop protection happens in checkExecutionLimit, not here
+			SplitInBatchesV3.resetExecutionCount(this);
 			return [nodeContext.processedItems, []];
 		}
 
 		nodeContext.done = false;
 
 		return [[], returnItems];
 	}
+
+	/**
+	 * Prevents infinite loops by limiting executions (PAY-2940)
+	 */
+	private static checkExecutionLimit(executeFunctions: IExecuteFunctions): void {
+		const nodeName = executeFunctions.getNode().name;
+		const executionId = executeFunctions.getExecutionId();
+		const globalKey = `${executionId}_${nodeName}`;
+
+		// Track execution timestamp for cleanup
+		SplitInBatchesV3.executionTracking.set(executionId, Date.now());
+
+		// Initialize and increment counter in static map
+		const currentCount = SplitInBatchesV3.executionCounters.get(globalKey) || 0;
+		const newCount = currentCount + 1;
+		SplitInBatchesV3.executionCounters.set(globalKey, newCount);
+
+		// Calculate dynamic limit based on input data and batch size
+		const inputItems = executeFunctions.getInputData();
+		const batchSize = executeFunctions.getNodeParameter('batchSize', 0) as number;
+
+		// Expected batches = ceil(inputItems / batchSize) + some buffer for edge cases
+		// Minimum of 10 to handle small datasets, maximum of 1000 to prevent true infinite loops
+		const expectedBatches = Math.ceil(inputItems.length / Math.max(batchSize, 1));
+		const bufferMultiplier = 2; // Allow 2x expected batches for safety
+		const maxExecutions = Math.min(Math.max(expectedBatches * bufferMultiplier, 10), 1000);
```

Comment:
Not sure if this will hold up in all cases, I could see cases where the done output is connected back to the input of the loop with an If node in between, and it could generate more executions than `maxExecutions` without being an infinite loop.